python/mxnet/gluon/data/dataset.py

```diff
@@ -64,6 +64,37 @@ def filter(self, fn):
         from . import FilterSampler
         return _SampledDataset(self, FilterSampler(fn, self))
 
+    def shard(self, num_shards, index):
+        """Returns a new dataset includes only 1/num_shards of this dataset.
+
+        For distributed training, be sure to shard before you randomize the dataset
+        (such as shuffle), if you want each worker to reach a unique subset.
+
+        Parameters
+        ----------
+        num_shards : int
+            A integer representing the number of data shards.
+        index : int
+            A integer representing the index of the current shard.
+
+        Returns
+        -------
+        Dataset
+            The result dataset.
+        """
+        assert index < num_shards, 'Shard index of out bound: %d out of %d'%(index, num_shards)
+        assert num_shards > 0, 'Number of shards must be greater than 0'
+        length = len(self)
+        shard_len = length // num_shards
+        # Compute the start index for this partition
+        start = shard_len * index
+        # Compute the end index for this partition
+        end = start + shard_len
+        if index == num_shards - 1:
+            end = length
```

Comment:
The sampler is not uniform. If there are 199 samples and 100 partitions, there will be `1+99` samples in the last partition, but `1` sample in other partitions.

The following implementation will be more uniform.
```python
shared_len = length // num_shareds
rest = length % num_shareds
start = shared_len * index + min(index, rest)
end = start + shared_len + (index < rest)
```